op-batcher/batcher/espresso/streamer.go

```diff
@@ -0,0 +1,232 @@
+package espresso
+
+import (
+	// #cgo darwin,arm64 LDFLAGS: -framework CoreFoundation -framework SystemConfiguration
+	"C"
+	"cmp"
+	"context"
+	"encoding/json"
+	"fmt"
+	"math/big"
+	"slices"
+
+	espressoClient "github.com/EspressoSystems/espresso-network-go/client"
+	espressoLightClient "github.com/EspressoSystems/espresso-network-go/light-client"
+	espressoTypes "github.com/EspressoSystems/espresso-network-go/types"
+	espressoVerification "github.com/EspressoSystems/espresso-network-go/verification"
+
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+// espresso-network-go's HeaderInterface currently lacks a function to get this info,
+// although it is present in all header versions
+func getFinalizedL1(header *espressoTypes.HeaderImpl) *espressoTypes.L1BlockInfo {
+	v0_1, ok := header.Header.(*espressoTypes.Header0_1)
+	if ok {
+		return v0_1.L1Finalized
+	}
+	v0_2, ok := header.Header.(*espressoTypes.Header0_2)
+	if ok {
+		return v0_2.L1Finalized
+	}
+	v0_3, ok := header.Header.(*espressoTypes.Header0_3)
+	if ok {
+		return v0_3.L1Finalized
+	}
+	return nil
+}
+
+type L1Client interface {
+	HeaderByNumber(ctx context.Context, number *big.Int) (*types.Header, error)
+}
+
+type EspressoStreamer struct {
+	// Namespace of the rollup we're interested in
+	Namespace uint64
+	// Address of the batcher, we expect transactions to
+	// be signed by the corresponding private key
+	BatcherAddress common.Address
+
+	L1Client            L1Client
+	EspressoClient      *espressoClient.Client
+	EspressoLightClient *espressoLightClient.LightClientReader
+	Log                 log.Logger
+
+	// Batch number we're to give out next
+	BatchPos uint64
+	// HotShot block we're to fetch next
+	hotShotPos uint64
+	// Position of the last safe batch
+	confirmedBatchPos uint64
+	// Hotshot block corresponding to the last safe batch
+	confirmedHotShotPos uint64
+
+	// Maintained in sorted order, but may be missing batches if we receive
+	// any out of order.
+	batchBuffer []EspressoBatch
```

Comment:
Does this buffer contains all the batches from the beginning or only since `confirmedBatchPos`?